include/deal.II/differentiation/sd/symengine_number_types.h

```diff
@@ -0,0 +1,1036 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2019 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_sd_symengine_number_types_h
+#define dealii_differentiation_sd_symengine_number_types_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_SYMENGINE
+
+// Low level
+#  include <symengine/basic.h>
+#  include <symengine/dict.h>
+#  include <symengine/symengine_exception.h>
+#  include <symengine/symengine_rcp.h>
+
+// Number types
+#  include <symengine/expression.h>
+#  include <symengine/integer.h>
+#  include <symengine/logic.h>
+#  include <symengine/rational.h>
+
+// Number operations
+#  include <symengine/add.h>
+#  include <symengine/functions.h>
+#  include <symengine/mul.h>
+#  include <symengine/pow.h>
+
+// Evaluation
+#  include <symengine/eval.h>
+#  include <symengine/eval_arb.h>
+#  include <symengine/eval_double.h>
+
+// Differentiation
+#  include <deal.II/base/exceptions.h>
+
+#  include <deal.II/differentiation/sd/symengine_number_traits.h>
+#  include <deal.II/differentiation/sd/symengine_types.h>
+
+#  include <symengine/derivative.h>
+
+#  include <algorithm>
+#  include <memory>
+#  include <type_traits>
+#  include <utility>
+#  include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace SD
+  {
+    namespace SE = ::SymEngine;
+
+    /**
+     * A class to wrap SymEngine expressions.
+     *
+     * With this number type, SymEngine numbers can be used to perform scalar
+     * and tensor mathematics in deal.II. It (or the SymEngine::Expression
+     * class, of which it stores an instance) therefore forms the basis of
+     * symbolic computation via SymEngine in deal.II. With it one can
+     * perform symbolic differentiation and subsequent substitution with both
+     * scalars and deal.II's native Tensor and SymmetricTensor types.
+     *
```

Comment:
Maybe add a small example here, on how to use this class?

Reply:
I've added a number of examples, which I'm sure should suffice :-)